OAPE-194: UPSTREAM: <carry>: Add downstream README file (#46)
openshift/ansible-operator-plugins: openshift/README.md

54d3fc84 updating dependencies to support k8s 1.32 (#142)
dc609070 (tag: v1.37.2, upstream/release-v1.37) prepping for v1.37.2 release (#140)
e7a55338 Fix CVE-2025-27516 by updating ansible-core to v2.18.3 (#139)
fac29c58 remove `--ignore 70612` from  pipenv check in pipfile.Dockerfile (#134)
c315031e prepping for v1.37.1 release (#128)
e2b748b2 fixing go and python cve (#126)
d88e9be0 (tag: v1.37.0) prepping for v1.37.0 release (#123)
83876cf1 updating to go 1.23 and UBI 9.5 (#122)
bef914f8 Update k8s dependencies to 1.31 (#107)
d23c8aa6 prepping for v1.36.1 release (#117)
3ad1b57a adding metrics-require-rbac flag, metrics-secure validation logic and corresponding scaffolding logic (#116)
3aaf1112 removing operator types from issues (#112)
[v1.38.0-rebase-main 62007ace] UPSTREAM: <drop>: Update vendor directory
 3474 files changed, 185246 insertions(+), 148228 deletions(-)
...
...
...
```

When the script is finished, you will be in a `UPSTREAM-TAG-rebase-main`
branch. In our example above we were in `v1.38.0-rebase-main`. For more
information about the output see the [breakdown of rebase_upstream.sh
output](#breakdown-of-rebase_upstreamsh-output) section below.

> At this point, verify things look okay. Ensure that the `./openshift/release/ansible/ansible_collections` directory

Reviewer: ```suggestion
At this point, you _must_ verify if everything looks okay. Ensure that the `./openshift/release/ansible/ansible_collections` directory
```

and the requirements files in the `./openshift` directory are properly generated. This will require manual intervention and
inspection.

Once these directories have been verified, create a release PR downstream:

```
git push origin v1.38.0-rebase-main
```

### Breakdown of rebase_upstream.sh output

Running `rebase_upstream.sh` will spew out a bunch of output. We'll break it down
here.

> The script will verify your branch is up to date and create a new branch

Reviewer: ```suggestion
The script will verify if your branch is up to date and create a new branch
```

3aaf1112 removing operator types from issues (#112)
...
```

After the commits are listed, the vendor directory is updata:
```
[v1.38.0-rebase-main 62007ace] UPSTREAM: <drop>: Update vendor directory
 3474 files changed, 185246 insertions(+), 148228 deletions(-)
...
...
...
```

Then, the `ansible_collections` directory and the downstream requirements files are updated:

```
[v1.38.0-rebase-main e3b0c442] UPSTREAM: <carry>: Update ansible_collections directory

[v1.38.0-rebase-main dcd0d1b9] UPSTREAM: <carry>: Update downstream requirements
```

Finally you will get a completed message:

```
** Upstream merge complete! **
View the above incoming commits to verify all is well
(mirrors the commit listing the PR will show)

Now make a pull request.
```

## Patch release to specific release branch

There are times when you will want to bring an upstream patch release to the
downstream OpenShift release. For example, bringing down v1.37.2 to downstream
release-4.17.

We will use the same script, `rebase_upstream.sh` except we will add the branch
we want to patch.

```
./openshift/hack/rebase_upstream.sh v1.37.2 release-4.17
```

Here is the output of running the above script.

```

> $ ./openshift/hack/rebase_upstream.sh v1.37.2 release-4.17

Reviewer: Please hide the output by default.

9871a1eb updating all non k8s go dependencies to latest (#108)
d5cde0cd Bump k8s dependencies to 1.30 (#102)
886b92ad Update github actions to use node 20 (#85)
3acad86d mark unsafe the non-spec obj parameter (#66)
f1aa5843 Updates ansible collections and ansible packages to supported versions (#105)
ab6893df Utilize GitHub Actions cache for image builds (#104)
9a6d15a1 Bump kubebuilder to 3.15.0 (with merge conflicts resolved) (#90)
8e08a453 Bump base image to RHEL9 and Python version to 3.12 (#101)
0818d9ab Bump base image to RHEL9 and Python version to 3.12
311f40b2 Fix deleted files
9dcddde9 bump kubebuilder to 3.15.0
36380ebe Update dependencies to resolve safety-cli alerts (#98)
3b42c75e Update dependencies to resolve safety-cli alerts
[v1.37.2-rebase-release-4.17 98007221] UPSTREAM: <drop>: Update vendor directory
 3224 files changed, 285459 insertions(+), 75421 deletions(-)
...
...
...
```

Just like sync to main, verify things look okay. Ensure that the `./openshift/release/ansible/ansible_collections` directory
and the requirements files in the `./openshift` directory are properly generated. This will require manual intervention and
inspection.

> Once these directories have been verified, create a release PR downstream:

Reviewer: ```suggestion
Once these directories have been verified, create a rebase PR downstream:
```

54d3fc84 updating dependencies to support k8s 1.32 (#142)
dc609070 (tag: v1.37.2, upstream/release-v1.37) prepping for v1.37.2 release (#140)
e7a55338 Fix CVE-2025-27516 by updating ansible-core to v2.18.3 (#139)
fac29c58 remove `--ignore 70612` from  pipenv check in pipfile.Dockerfile (#134)
c315031e prepping for v1.37.1 release (#128)
e2b748b2 fixing go and python cve (#126)
d88e9be0 (tag: v1.37.0) prepping for v1.37.0 release (#123)
83876cf1 updating to go 1.23 and UBI 9.5 (#122)
bef914f8 Update k8s dependencies to 1.31 (#107)
d23c8aa6 prepping for v1.36.1 release (#117)
3ad1b57a adding metrics-require-rbac flag, metrics-secure validation logic and corresponding scaffolding logic (#116)
3aaf1112 removing operator types from issues (#112)
[v1.38.0-rebase-main 62007ace] UPSTREAM: <drop>: Update vendor directory
 3474 files changed, 185246 insertions(+), 148228 deletions(-)
...
...
...
```
</details>

When the script is finished, you will be in a `UPSTREAM-TAG-rebase-main`
branch. In our example above we were in `v1.38.0-rebase-main`. For more
information about the output see the [breakdown of rebase_upstream.sh
output](#breakdown-of-rebase_upstreamsh-output) section below.

> At this point, you _must_ verify if everything looks. Ensure that the `./openshift/release/ansible/ansible_collections` directory

Reviewer: ```suggestion
At this point, you _must_ verify if everything looks okay. Ensure that the `./openshift/release/ansible/ansible_collections` directory
```
